src/providers/aws/mod.rs

```diff
@@ -49,29 +50,92 @@ pub struct AwsProvider {
 
 impl AwsProvider {
     pub fn try_new() -> Result<AwsProvider> {
-        let client = retry::Client::try_new()?.return_on_404(true);
+        let mut client = retry::Client::try_new()?.return_on_404(true);
+        let token = AwsProvider::fetch_imdsv2_token(client.clone());
+
+        // If IMDSv2 token is fetched successfully, set the header.
+        // Otherwise, proceed with IMDSv1 mechanism.
+        match token {
+            Ok(t) => {
+                client = client.header(
+                    header::HeaderName::from_bytes(b"X-aws-ec2-metadata-token")
+                        .chain_err(|| "setting header name for aws imdsv2 metadata")?,
+                    header::HeaderValue::from_bytes(t.as_bytes())
+                        .chain_err(|| "setting header value for aws imdsv2 metadata")?,
+                );
+            }
+            Err(err) => {
+                warn!("failed to fetch aws imdsv2 session token with: {}", err);
+            }
+        }
+
+        Ok(AwsProvider { client })
+    }
+
+    fn _try_new_with_client(mut client: retry::Client) -> Result<AwsProvider> {
```

Comment:
Nit: `with_client` is probably enough as a name for a private method. Alternatively, `try_from_client` is ok too.

Also, you can drop the `mut` from the argument signature and rebind internally (because this function gets complete ownership of the client in input):
```rust
fn with_client(client: retry::Client) -> Result<AwsProvider> {
  let mut client = client;
  ...
}
```